File: src/main/java/com/philhanna/flashcards/port/in/StudySessionUseCase.java
package com.philhanna.flashcards.port.in;

import com.philhanna.flashcards.domain.session.Session;

/**
 * Inbound port: use cases for studying a flashcard deck.
 * The UI calls these methods; domain logic lives in the implementing service.
 */
public interface StudySessionUseCase {

   /** Marks the current card correct and advances to the next card */
   void markRight();

   /** Marks the current card incorrect and re-queues it */
   void markWrong();

   /** Advances to the next card */
   void nextCard();

   /** Goes back to the previous card */
   void previousCard();

   /** Returns the underlying session for read-only UI queries */
   Session getSession();
}

File: src/flashcards/port/study_session_use_case.py
# flashcards.port.study_session_use_case

from abc import ABC, abstractmethod

from flashcards.domain.session.session import Session


class StudySessionUseCase(ABC):

    @abstractmethod
    def mark_right(self) -> None: ...

    @abstractmethod
    def mark_wrong(self) -> None: ...

    @abstractmethod
    def next_card(self) -> None: ...

    @abstractmethod
    def previous_card(self) -> None: ...

    @property
    @abstractmethod
    def session(self) -> Session: ...
